### PR TITLE
{,nixos/}rkvm: init

### LINKED ISSUE #108276: package request: rkvm: Virtual KVM switch for Linux machines
**Project description**
It's like synergy or barrier, but supports wayland and windows as well.

I started packaging it [here](https://github.com/colemickens/nixcfg/tree/main/pkgs/rkvm), but it doesn't build yet (`build.rs` fails to find the evdev files it wants).

**Metadata**

* homepage URL: https://github.com/htrefil/rkvm
* source URL: https://github.com/htrefil/rkvm
* license: mit
* platforms: linux

### PR DESCRIPTION
**Pull request by ckiee**
## Description of changes

https://github.com/htrefil/rkvm: Virtual KVM switch for Linux machines

Fixes https://github.com/NixOS/nixpkgs/issues/108276

#### TODO
- [ ] hardening
- [ ] `passwordFile`
- [ ] autogenerate cert/key pair, `StateDirectory`
- [X] update to rkvm 0.5.1 when it comes out (current ver is a bit broken
- [x] nixos test (thanks @h7x4)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ckiee** [comment]
Updated rkvm to 0.4.1 (removing the patch, it got merged) and improved a link in the NixOS module.

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:16]: for consistency's sake
```suggestion
      description = mdDoc ''
```

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:30]: Settings for server and client are almost the same.  It would be great if we could avoid this redundancy.  [The frp module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/frp.nix) might be an example.

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:21]: Could you use `mkPackageOption` for simplicity?

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:145]: emm, why are systemd services from the package included but not used?  They are almost entirely overridden by the following `systemd.services`.

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:153]: The server and client services have different dependencies for network(-online).target.

Check my upstream fix [PR](https://github.com/htrefil/rkvm/pull/48) for more info.

File: pkgs/tools/misc/rkvm/default.nix
Comment (jian-lin) [inline pkgs/tools/misc/rkvm/default.nix:7]: These are not used.
```suggestion
```

File: pkgs/tools/misc/rkvm/default.nix
Comment (jian-lin) [inline pkgs/tools/misc/rkvm/default.nix:37]: nitpicking
```suggestion
    description = "A virtual KVM switch for Linux machines";
```

**Review by jian-lin — COMMENTED** [review]
An improvement is to harden the services, at least use `DynamicUser`.  This is doable.  Check [the kanata module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/hardware/kanata.nix) for an example.

This suggestion is not a blocker for this PR.

File: pkgs/tools/misc/rkvm/default.nix
Comment (h7x4) [inline pkgs/tools/misc/rkvm/default.nix:38]: Just an extra attribute.

```suggestion
    homepage = "https://github.com/htrefil/rkvm";
    changelog = "https://github.com/htrefil/rkvm/releases/tag/${version}";
```

**Review by h7x4 — CHANGES_REQUESTED** [review]

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:3]: Why? ripgrep in Doom shows 23139 usages of `with lib;` in nixpkgs:<br>
screenshot

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:43]: This was plain wrong, now corrected to what rkvm *actually* supports.

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:119]: oops!

**Comment by ckiee** [comment]
i'm not done going through your review yet, but sorry about all the obvious mistakes! i was very distracted while writing the NixOS module so i wasn't really thinking things through.

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:149]: This provides access to the binaries for the user's convenience, including the certificate generator.

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:157]: very nice catch! thank you (:

File: pkgs/tools/misc/rkvm/default.nix
Comment (ckiee) [reply pkgs/tools/misc/rkvm/default.nix:37]: I think the _"A "_ is gunk when you're trying to read a `nix search` to find something suitable (:

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:31]: This was an intentional choice, it allows for better docs and a smaller diff if one component gets more unshared options.

File: nixos/modules/services/misc/rkvm.nix
Comment (ckiee) [reply nixos/modules/services/misc/rkvm.nix:17]: thanks!

File: nixos/modules/services/misc/rkvm.nix
Comment (h7x4) [reply nixos/modules/services/misc/rkvm.nix:52]: I'll have a look to see if I can find the details.

As for autogenerating, I think it might bring a bit of extra complexity. The certificates requires you to specify the ip address (or a list of them) that the server will listen on, and that doesn't seem to go hand in hand with the `0.0.0.0` default. If we make a list option for this, should the certificate be regenerated if it gets changed?

File: nixos/modules/services/misc/rkvm.nix
Comment (h7x4) [reply nixos/modules/services/misc/rkvm.nix:85]: I'll just quickly reopen this in case you forgot

File: nixos/modules/services/misc/rkvm.nix
Comment (h7x4) [reply nixos/modules/services/misc/rkvm.nix:3]: I believe there are ongoing efforts to remove most top-level `with lib;` statements, due to some language intricacies. There's also this page: https://nix.dev/recipes/best-practices#with-scopes, which uses `pkgs` as an example but lists out some of the concerns.

I think the most maintainable solution is to use an `inherit (lib) ...` at the top for the most used stuff (like `mkOption` and `types`), and just prepend `lib.` for every function you only use once or twice. That way, there's no need to maintain a big list of everything used up at the top, while still keeping most of the code relatively clean.

Output of `git log --grep "with lib"` shows some of the efforts:

image

File: nixos/modules/services/misc/rkvm.nix
Comment (h7x4) [reply nixos/modules/services/misc/rkvm.nix]: Nice!

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [reply nixos/modules/services/misc/rkvm.nix:141]: > why are systemd services from the package included but not used? They are almost entirely overridden by the following systemd.services.

I misread `environment.systemPackages` as `systemd.packages`.

Should we use the upstream systemd files by `systemd.packages = [ cfg.package ]`?

> This provides access to the binaries for the user's convenience, including the certificate generator.

Makes senses.

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:152]: `network.target` should not be wanted by this service.  See `man systemd.special`.

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:132]: It seems that we have to use an ip address instead of a domain name.

File: nixos/modules/services/misc/rkvm.nix
Comment (jian-lin) [inline nixos/modules/services/misc/rkvm.nix:92]: This `rec` seems unneeded.

```suggestion
      settings = mkOption {
```

**Comment by ckiee** [comment]
feel free to commit your review fixes btw, I can squash before we merge if you don't want to force push, but otherwise it's unreasonably taking up hours of on-and-off nitpick fixes that do nearly nothing.

**Comment by jian-lin** [comment]
Could you clarify what comments you refer to as nitpicking?

Since you make changes for two of my three new comments, it is likely that you think [the comment](https://github.com/NixOS/nixpkgs/pull/250792#discussion_r1346159368) about `network.target` is nitpicking.  No, it is not.  I regard that as an error since `man systemd.special` says clearly that what you do is wrong.  It may works well but it is still wrong.

> it's unreasonably taking up hours

I do not understand what takes you hours to do since the changes you made according to reviews clearly does not need that much time.

I understand that nitpicking comments can be tiring.  So I seldom give them and when I do leave one, I say clearly that it is nitpicking.

**Comment by ckiee** [comment]
nixpkgs is a volunteer project and I'm not currently able to volunteer, so feel free to commit the changes you want yourself; no objections there. I've already had rkvm working on my NixOS systems working for a month, and [the code](https://github.com/ckiee/nixfiles/tree/09df5089ff6caada1f8e715da3dbf9562363f6a4/modules/rkvm.nix) is freely available which is more than enough.

Nothing more was implied, and regaining focus after addressing the latest comments takes a while — maybe longer for me than for you. Thank you for your understanding (:

**Review by jian-lin — APPROVED** [review]
LGTM

I test it locally and it works well.

Secret-handling and hardening can be done in another PR.

**Comment by ckiee** [comment]
@ofborg test rkvm

**Comment by jian-lin** [comment]
It is weird that `nixosTests.rkvm` on aarch64-linux fails but `rkvm, rkvm.passthru.tests` on aarch64-linux succeeds.

@ofborg test rkvm

File: pkgs/tools/misc/rkvm/default.nix
Comment (Scrumplex) [inline pkgs/tools/misc/rkvm/default.nix:46]: This should have a `mainProgram` attribute

File: pkgs/top-level/all-packages.nix
Comment (Scrumplex) [inline pkgs/top-level/all-packages.nix:12449]: Maybe use `pkgs/by-name`? See [RFC 0140](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md)

File: pkgs/tools/misc/rkvm/default.nix
Comment (jian-lin) [reply pkgs/tools/misc/rkvm/default.nix:46]: I do not think so.  It has three binaries, and none of them should be the main one.

File: pkgs/top-level/all-packages.nix
Comment (jian-lin) [reply pkgs/top-level/all-packages.nix:12449]: This should not be a blocker.  IIRC, there will be tree-wide PR for these things.

File: pkgs/tools/misc/rkvm/default.nix
Comment (Scrumplex) [reply pkgs/tools/misc/rkvm/default.nix:46]: Hmm. Not quite sure. There are lots of other multi-binary packages where there isn't a definitive "main" binary. I.e. `redis` chooses `redis-client` as its main binary.

The README of rkvm mentions this:

> - Since rkvm-server grabs all input, i's a good idea to do a test run first to make sure you won't end up
>   being unable to user your keyboard and/or mouse because your display server is not properly configured to receive input from rkvm.
> 
>   Run the following command to start rkvm-server for 15 seconds to test that your keyboard, mouse, etc. works properly:
>   ```
>   # rkvm-server /etc/rkvm/server.toml --shutdown-after 15
>   ```

So perhaps the server should be the main binary?

Comment (jian-lin) [reply pkgs/tools/misc/rkvm/default.nix:46]: What about adding mainProgram when a real user case occurs?

Since the author is [busy](https://github.com/NixOS/nixpkgs/pull/250792#issuecomment-1747351363) and this PR is in a good state, I suggest we merge it as is for now.

**Comment by jian-lin** [comment]
@ofborg test rkvm

File: nixos/modules/services/misc/rkvm.nix
Comment (h7x4) [reply nixos/modules/services/misc/rkvm.nix:3]: I'm not going to block this PR on this issue, but I'll link #208242. I don't really mind either outcome, but it would be good to come to some sort of agreement, so people don't have to go back and forth on this for every PR.

**Comment by ckiee** [comment]
now running your bump on my system, @jian-lin. i'll rebase soon (probably tomorrow) if it seems to work well.

**Comment by ckiee** [comment]
it works very well! 0.5.1 seems to be much more stable (:

i think we are a go for merge, if there are no more objections?

**Review by Tungsten842 — APPROVED** [review]
It works well

**Comment by ckiee** [comment]
thanks everybody! i'm glad to see this in & through (:

**Pull request merged**
